src/js/components/modals/ServiceFormModal.js

```diff
@@ -0,0 +1,240 @@
+import Ace from 'react-ace';
+import mixin from 'reactjs-mixin';
+import {Modal} from 'reactjs-components';
+import React from 'react';
+import {StoreMixin} from 'mesosphere-shared-reactjs';
+
+import 'brace/mode/json';
+import 'brace/theme/monokai';
+import 'brace/ext/language_tools';
+
+import MarathonStore from '../../stores/MarathonStore';
+import SchemaForm from '../SchemaForm';
+import Service from '../../structs/Service';
+import ServiceUtil from '../../utils/ServiceUtil';
+import ServiceSchema from '../../constants/ServiceSchema';
+import ToggleButton from '../ToggleButton';
+
+const METHODS_TO_BIND = [
+  'getAdvancedSubmit',
+  'handleCancel',
+  'handleClearError',
+  'handleJSONChange',
+  'handleJSONToggle',
+  'handleSubmit',
+  'onMarathonStoreServiceCreateError',
+  'onMarathonStoreServiceCreateSuccess'
+];
+
+class ServiceFormModal extends mixin(StoreMixin) {
+  constructor() {
+    super(...arguments);
+
+    let model =
+      ServiceUtil.createFormModelFromSchema(ServiceSchema);
+    this.state = {
+      app: JSON.stringify({id:'', cmd:''}, null, 2),
+      errorMessage: null,
+      jsonMode: false,
+      model,
+      service: ServiceUtil.createServiceFromFormModel(model)
+    };
+
+    this.store_listeners = [
+      {
+        name: 'marathon',
+        events: ['serviceCreateError', 'serviceCreateSuccess']
+      }
+    ];
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  componentWillReceiveProps(nextProps) {
+    super.componentWillReceiveProps(...arguments);
+    if (!this.props.open && nextProps.open) {
+      this.resetState();
+    }
+  }
+
+  resetState() {
+    let model = ServiceUtil.createFormModelFromSchema(ServiceSchema);
+    this.setState({
+      app: '',
+      errorMessage: null,
+      jsonMode: false,
+      model,
+      service: ServiceUtil.createServiceFromFormModel(model)
+    });
+  }
+
+  handleClearError() {
+    this.setState({
+      errorMessage: null
+    });
+  }
+
+  handleJSONChange(app) {
+    let {service} = this.state;
+    try {
+      service = new Service(JSON.parse(app));
+    } catch (e) {
+
+    }
+    this.setState({app, service})
+  }
+
+  handleJSONToggle() {
+    let nextState = {};
+    if (!this.state.jsonMode) {
+      let {model} = this.triggerAdvancedSubmit();
+      let service = ServiceUtil.createServiceFromFormModel(model);
+      nextState.model = model;
+      nextState.service = service;
+      nextState.app = JSON.stringify(ServiceUtil
+        .getAppDefinitionFromService(service), null, 2);
+    }
+    nextState.jsonMode = !this.state.jsonMode;
+    this.setState(nextState);
+  }
+
+  onMarathonStoreServiceCreateSuccess() {
+    this.resetState();
+    this.props.onClose();
+  }
+
+  onMarathonStoreServiceCreateError(errorMessage) {
+    this.setState({
+      errorMessage
+    });
+  }
+
+  handleCancel() {
+    this.props.onClose();
```

Comment:
`props.onClose` is optional, let's either make it required or ensure we don't try to call `undefined`().